subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/properties/DefaultPropertyWalkerTest.groovy

```diff
@@ -111,6 +111,17 @@ class DefaultPropertyWalkerTest extends AbstractProjectBuilderSpec {
         1 * visitor.visitInputProperty({ it.propertyName == 'bean' && it.value == null })
     }
 
+    @Unroll
+    def "correct implementation for #type coerced to Action is tracked"() {
+        expect:
+        DefaultPropertyWalker.getImplementationClass(implementation as Action) == implementation.getClass()
+
+        where:
+        type      | implementation
+        "Closure" | { it }
+        "Action"  |  new Action<String>() { @Override void execute(String s) {} }
+    }
```

Comment:
Actually, I think one `SAM` type is enough. I don't see why other SAM types would give better coverage for Groovy.
For Java and Kotlin Lambdas we would need a test written in Java and one in Kotlin. I didn't add those since the test is written in Groovy and we already have integration tests for Kotlin and Java lambdas.
Sounds good?

Reply:
Makes sense.